## Logistic Regression


Logistic Regression is a statistical method used for binary classification problems, where the outcome can take one of two possible values (e.g., yes/no, true/false, 0/1). It models the probability that a given input belongs to a particular class.

### Key Concepts

#### Logistic Function (Sigmoid Function)
The logistic regression model uses the logistic function to map predicted values to probabilities. The logistic function is defined as:

\[ \sigma(z) = \frac{1}{1 + e^{-z}} \]

where \( z \) is the input to the function, which is a linear combination of the input features.

#### Model Representation
The hypothesis of logistic regression is given by:

\[ h_\theta(x) = \sigma(\theta^T x) = \frac{1}{1 + e^{-\theta^T x}} \]

where \( \theta \) is the vector of parameters (weights) and \( x \) is the vector of input features.

#### Cost Function
The cost function for logistic regression is the logistic loss (also known as log loss or cross-entropy loss):

\[ J(\theta) = - \frac{1}{m} \sum_{i=1}^m \left[ y^{(i)} \log(h_\theta(x^{(i)})) + (1 - y^{(i)}) \log(1 - h_\theta(x^{(i)})) \right] \]

where \( m \) is the number of training examples, \( y^{(i)} \) is the true label, and \( h_\theta(x^{(i)})) \) is the predicted probability for the \(i\)-th example.

#### Training the Model
The model is trained using optimization algorithms like gradient descent to minimize the cost function and find the optimal parameters \( \theta \).

### Example Code Using Python and Scikit-Learn

Here's an example of how to implement logistic regression using Python's Scikit-Learn library:

```python
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import datasets
import matplotlib.pyplot as plt

# Load a dataset (e.g., iris dataset for simplicity)
iris = datasets.load_iris()
X = iris.data[:, :2]  # Use only the first two features for visualization
y = (iris.target != 0) * 1  # Convert to a binary classification problem (e.g., class 0 vs. not class 0)

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Create and train the logistic regression model
model = LogisticRegression()
model.fit(X_train, y_train)

# Predict the labels for the test set
y_pred = model.predict(X_test)

# Calculate accuracy
accuracy = model.score(X_test, y_test)
print(f'Accuracy: {accuracy:.2f}')

# Plot decision boundary
x_min, x_max = X[:, 0].min() - 0.5, X[:, 0].max() + 0.5
y_min, y_max = X[:, 1].min() - 0.5, X[:, 1].max() + 0.5
xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.01), np.arange(y_min, y_max, 0.01))
Z = model.predict(np.c_[xx.ravel(), yy.ravel()])
Z = Z.reshape(xx.shape)

plt.contourf(xx, yy, Z, alpha=0.8)
plt.scatter(X[:, 0], X[:, 1], c=y, edgecolors='k', marker='o')
plt.xlabel(iris.feature_names[0])
plt.ylabel(iris.feature_names[1])
plt.title('Logistic Regression Decision Boundary')
plt.show()
```

### Key Points to Remember
- **Binary Classification**: Logistic regression is primarily used for binary classification problems.
- **Probability Output**: The model outputs probabilities that are mapped to class labels using a threshold (e.g., 0.5).
- **Cost Function**: The log loss function is used to measure the performance of the model.
- **Interpretability**: The coefficients \( \theta \) in the logistic regression model can be interpreted to understand the influence of each feature on the probability of the outcome.

## Coefficient
- In logistic regression the coefficient is the expected change in log-odds of having the outcome per unit change in X.
- This does not have the most intuitive understanding so let's use it to create something that makes more sense, odds

In [1]:
import numpy
from sklearn import linear_model

#Reshaped for Logistic function.
X = numpy.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = numpy.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

log_odds = logr.coef_
odds = numpy.exp(log_odds)

print(odds)

[[4.03557295]]


## Probability

In [2]:
import numpy
from sklearn import linear_model

X = numpy.array([3.78, 2.44, 2.09, 0.14, 1.72, 1.65, 4.92, 4.37, 4.96, 4.52, 3.69, 5.88]).reshape(-1,1)
y = numpy.array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1])

logr = linear_model.LogisticRegression()
logr.fit(X,y)

def logit2prob(logr, X):
  log_odds = logr.coef_ * X + logr.intercept_
  odds = numpy.exp(log_odds)
  probability = odds / (1 + odds)
  return(probability)

print(logit2prob(logr, X))

[[0.60749168]
 [0.19267555]
 [0.12774788]
 [0.00955056]
 [0.08037781]
 [0.0734485 ]
 [0.88362857]
 [0.77901203]
 [0.88924534]
 [0.81293431]
 [0.57718238]
 [0.96664398]]


3.78 0.61 The probability that a tumor with the size 3.78cm is cancerous is 61%.

2.44 0.19 The probability that a tumor with the size 2.44cm is cancerous is 19%.

2.09 0.13 The probability that a tumor with the size 2.09cm is cancerous is 13%.

